core/src/main/java/com/sequenceiq/distrox/v1/distrox/service/upgrade/DistroxUpgradeImageSelector.java

```diff
@@ -0,0 +1,88 @@
+package com.sequenceiq.distrox.v1.distrox.service.upgrade;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+import com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.request.tags.upgrade.UpgradeV4Request;
+import com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.response.image.ImageComponentVersions;
+import com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.response.image.ImageInfoV4Response;
+import com.sequenceiq.cloudbreak.exception.BadRequestException;
+
+@Component
+public class DistroxUpgradeImageSelector {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DistroxUpgradeImageSelector.class);
+
+    public ImageInfoV4Response determineImageId(UpgradeV4Request request, List<ImageInfoV4Response> upgradeCandidates) {
+        Optional<ImageInfoV4Response> image = Optional.empty();
+        if (Objects.isNull(request) || request.isEmpty() || Boolean.TRUE.equals(request.getLockComponents())) {
+            image = selectLatestImageFromCandidates(upgradeCandidates);
+        } else if (StringUtils.isNotEmpty(request.getImageId())) {
+            image = selectRequestedImage(request, upgradeCandidates);
+        } else if (StringUtils.isNotEmpty(request.getRuntime())) {
+            image = selectLastestImageByRuntime(upgradeCandidates, request.getRuntime());
+        }
+        return image.orElseThrow(() -> new BadRequestException(String.format("Invalid upgrade request, please validate the contents: %s", request)));
+    }
+
+    private Optional<ImageInfoV4Response> selectLastestImageByRuntime(List<ImageInfoV4Response> upgradeCandidates, String runtime) {
+        List<ImageInfoV4Response> imagesWithMatchingRuntime = upgradeCandidates.stream().filter(
+                imageInfoV4Response -> runtime.equals(imageInfoV4Response.getComponentVersions().getCdp())).collect(Collectors.toList());
+        validateThereIsMatchingRuntime(upgradeCandidates, imagesWithMatchingRuntime, runtime);
+        ImageInfoV4Response imageInfoV4Response = imagesWithMatchingRuntime.stream().max(ImageInfoV4Response.creationBasedComparator()).orElseThrow();
```

Comment:
this should be instead:         Optional<ImageInfoV4Response> imageInfoV4Response = imagesWithMatchingRuntime.stream().max(ImageInfoV4Response.creationBasedComparator());